flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskTest.java

```diff
@@ -1727,6 +1731,99 @@ protected void cancelTask() {
         assertTrue(OpenFailingOperator.wasClosed);
     }
 
+    @Test
+    public void testMailboxNotifyThroughputMeterAboutIdleTime() throws Exception {
+        final long sleepTimeOutsideMail = 42;
+        final long sleepTimeInsideMail = 44;
+        int incomingDataSize = 10_000;
+
+        @Nullable WaitingThread waitingThread = null;
+        try (final MockEnvironment environment = setupEnvironment(true, true)) {
+            final UnAvailableTestInputProcessor inputProcessor =
+                    new UnAvailableTestInputProcessor();
+            final StreamTask task =
+                    new MockStreamTaskBuilder(environment)
+                            .setStreamInputProcessor(inputProcessor)
+                            .build();
+            TaskIOMetricGroup ioMetricGroup =
+                    task.getEnvironment().getMetricGroup().getIOMetricGroup();
+            ThroughputCalculator throughputCalculator = environment.getThroughputMeter();
+
+            final MailboxExecutor executor = task.mailboxProcessor.getMainMailboxExecutor();
+            final RunnableWithException completeFutureTask =
+                    () ->
+                            inputProcessor
+                                    .availabilityProvider
+                                    .getUnavailableToResetAvailable()
+                                    .complete(null);
+
+            waitingThread =
+                    new WaitingThread(
+                            executor,
+                            completeFutureTask,
+                            sleepTimeInsideMail,
+                            sleepTimeOutsideMail,
+                            ioMetricGroup.getIdleTimeMsPerSecond());
+            // Make sure WaitingThread is started after Task starts processing.
+            executor.submit(
+                    waitingThread::start,
+                    "Start WaitingThread after Task starts processing input.");
+
+            long startTs = System.currentTimeMillis();
+            throughputCalculator.calculateThroughput();
+            throughputCalculator.incomingDataSize(incomingDataSize);
+            task.invoke();
+            long resultThroughput = throughputCalculator.calculateThroughput();
+            long totalDuration = System.currentTimeMillis() - startTs;
+
+            assertThat(
+                    resultThroughput,
+                    greaterThanOrEqualTo(
+                            incomingDataSize * 1000 / (totalDuration - sleepTimeOutsideMail)));
```

Comment:
This test duplicates quite a lot of code from `testProcessWithUnAvailableInput`. I would either:
1. put those highlighted lines inside `testProcessWithUnAvailableInput`
2. or deduplicate in another way the test setup 

I'm not sure how 2. could be done easily, so I would go with 1?